ZEP008-hot-backup.md

```diff
@@ -0,0 +1,659 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In this section we describe the reasoning behind the high-level algorithm from guide-level explanation. Then we extend the high-level algorithm to allow asynchronous backups and discuss the failure scenarios and edge cases.
+
+### What is a consistent backup
+
+A global checkpoint (backup) in a distributed system consists of a set of local checkpoints from the participating nodes. A *consistent* global checkpoint is a set of local checkpoints from which the nodes can recover without any inconsistencies in the state of each node in relation to the state of any other node.
+
+In a few words a global checkpoint is consistent if it is a consistent cut.
+
+To understand a consistent cut, let us introduce the concept of happened before relation.
+
+An event $e_i^h$ occurred in process i at logical time h. A happened before ($\rightarrow$) relation between two events is defined as follows:  
+An event $e_i^h \rightarrow e_j^k$  if and only if
+ $(i = j \wedge k = h + 1) \vee$
+ $(e_i^h = send(m) \wedge e_j^k = receive(m)) \vee$
+ $(\exists e_p^n : (e_i^h \rightarrow e_p^n) \wedge ( e_p^n \rightarrow) e_j^k))$
+
+That means an event e1 happens before e2 iff:
+- both events happened in the same process and e1 occurred before e2
+- e1 sends a message and e2 receives it
+- e1 happened before e2 transitively
+
+A cut is a subset of events in the state.
+A cut C is consistent if : $\forall e_i, e_j : e_j \in C \wedge e_i \rightarrow e_j \implies e_i \in C$.
+That means in consistent cut C, if an event exists in C then all events *happened before* it must also exist in C.
+
+
+![consistent-cut](ZEP008/consistent-cut.png)
+
+In the above image, the horizontal line depicts the time in each process. The vertical slanted arrow shows communication from one process to another. Dotted lines represent a cut. That means - we take a checkpoint where the dotted line meets the horizontal timeline. The green ones are a consistent cut. The red lines are cuts, but they are not consistent cuts. The red line is not consistent because e2 is in the cut, but e1 is not. e1 has happened before e2 because it is a "send-receive" pair.
+
+### Algorithm to find a consistent cut
+
+[BCS algorithm as proposed in this paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.149.5565&rep=rep1&type=pdf) is a way to take coordinated consistent checkpoints across multiple processes. (An easy to understand version is [here](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.78.4041&rep=rep1&type=pdf)). This is a communication induced checkpointing and relies on lazy coordination among the participating processes.
+
+Here is how the algorithm works.
+
+$C_i^{sn_i}$ is a local checkpoint of process $i$, and its checkpoint id is ${sn_i}$. When a process sends a message, it also embeds its current checkpoint id ${sn_i}$ with it.
```

Comment:
:x: The formatting is broken here for some reason. Here's what I see:

![image](https://user-images.githubusercontent.com/43373/175011095-a2aa04e3-2933-47ad-8bec-4ce09e07abf4.png)